src/bors/merge_queue.rs

```diff
@@ -0,0 +1,220 @@
+                    // We can only have one PR being built at a time - block the queue.
+                    break;
+                }
+                Ok(false) => {
+                    // Failed due to issue with the PR (e.g. merge conflicts).
+                    tracing::debug!("Failed to start merge build for PR {pr_num}");
+                    continue;
+                }
+                Err(error) => {
+                    // Unexpected error - the PR will remain in the "queue" for a retry.
+                    tracing::error!("Error starting merge build for PR {pr_num}: {:?}", error);
+                    continue;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Starts a new auto build for a pull request.
+async fn start_auto_build(
+    repo: &Arc<RepositoryState>,
+    ctx: &Arc<BorsContext>,
+    pr: PullRequestModel,
+) -> anyhow::Result<bool> {
+    let client = &repo.client;
+
+    let gh_pr = client.get_pull_request(pr.number).await?;
+    let base_sha = client.get_branch_sha(&pr.base_branch).await?;
+
+    let auto_merge_commit_message = format!(
+        "Auto merge of #{} - {}, r={}\n\n{}\n\n{}",
+        pr.number,
+        gh_pr.head_label,
+        pr.approver().unwrap_or("<unknown>"),
+        pr.title,
+        gh_pr.message
+    );
+
+    // 1. Merge PR head with base branch on `AUTO_MERGE_BRANCH_NAME`
+    match attempt_merge(
+        &repo.client,
+        &gh_pr.head.sha,
+        &base_sha,
+        &auto_merge_commit_message,
+    )
+    .await?
+    {
+        MergeResult::Success(merge_sha) => {
+            // 2. Push merge commit to `AUTO_BRANCH_NAME` where CI runs
+            client
+                .set_branch_to_sha(AUTO_BRANCH_NAME, &merge_sha)
+                .await?;
+
+            // 3. Record the build in the database
+            ctx.db
+                .attach_try_build(
+                    &pr,
+                    AUTO_BRANCH_NAME.to_string(),
+                    merge_sha.clone(),
+                    base_sha,
+                )
+                .await?;
+
+            // 4. Update label
+            handle_label_trigger(repo, pr.number, LabelTrigger::AutoBuildStarted).await?;
+
+            // 5. Post status comment
+            let comment = auto_build_started_comment(&gh_pr.head.sha, &merge_sha);
+            client.post_comment(pr.number, comment).await?;
+
+            // 6. Set GitHub commit status to pending on PR head
+            let desc = format!(
+                "Testing commit {} with merge {}...",
+                gh_pr.head.sha, merge_sha
+            );
+            client
+                .create_commit_status(
+                    &gh_pr.head.sha,
+                    StatusState::Pending,
+                    None,
+                    Some(&desc),
+                    Some("bors"),
+                )
+                .await?;
+
+            Ok(true)
+        }
+        MergeResult::Conflict => {
+            repo.client
```

Comment:
We might want to manually set the state of the PR in DB as "having conflicts" when this happens.